python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -2449,9 +2485,38 @@ def transfer_from(
 
                     # Rely on source_records being a defaultdict.
                     source_records[missing].extend(dataset_records)
+                return source_records
+
+            if True:
+                executor = concurrent.futures.ThreadPoolExecutor(max_workers=20)
+                future_chunks = [
+                    executor.submit(
+                        _check_missing_files, source_datastore, id_to_ref, artifact_existence, chunk
+                    )
+                    for chunk in chunk_iterable(missing_ids, chunk_size=5_000)
+                ]
+                for future in concurrent.futures.as_completed(future_chunks):
+                    # Let it fail if any chunk fails.
+                    new_records = future.result()
+                    for k, v in new_records.items():
+                        source_records[k].extend(v)
+            else:
+                for missing_ids_chunk in chunk_iterable(missing_ids, chunk_size=5_000):
+                    new_records = _check_missing_files(source_datastore, missing_ids_chunk)
+                    for k, v in new_records.items():
+                        source_records[k].extend(v)
+
+            log.verbose("Completed scan for missing data files")
 
         # See if we already have these records
-        target_records = self._get_stored_records_associated_with_refs(refs, ignore_datastore_records=True)
+        if not dry_run:
+            target_records = self._get_stored_records_associated_with_refs(
```

Comment:
This is the line I'm not sure about. I imagine people expect dry-run to know which records are really needed to be transferred rather than having the system think that it has to do all the checks again because the target doesn't have the datastore entries even when it does.

Reply:
Is this here because we're worried that the datasets might be present in the registry database, but not have file artifacts in the datastore and/or datastore records in the database?

I think it ought to be fine to assume that if a ref in the graph is in the registry we can skip it, as if it was inserted into the registry without associated datastore content that was probably intentional.

Reply:
I got rid of the change. The point was to try to make my profiles take as long as if the files weren't already transferred, but in reality it was a minor blip and it's better if the dry-run implementation does report that all the datasets have already been transferred and so you know that it won't do anything if you run it for real.